File: apps/watchkeeper/internal/workers/notifications.go
package workers

import (
	"bytes"
	"database/sql"
	"encoding/json"
	"errors"
	"fmt"
	"log"
	"net/http"
	"net/url"
	"os"
	"strings"
	"sync"
	"time"

	"github.com/stripsior/struxa/watchkeeper/internal/crypto"
)

const notifyHTTPTimeout = 8 * time.Second

type notifConfig struct {
	appName           string
	appURL            string
	discordEnabled    bool
	discordURL        string
	telegramEnabled   bool
	telegramToken     string
	telegramChatID    string
	userConfigEnabled bool
}

func loadNotifConfig(db *sql.DB) (notifConfig, error) {
	rows, err := db.Query("SELECT `key`, `value` FROM settings")
	if err != nil {
		return notifConfig{}, err
	}
	defer rows.Close()
	s := map[string]string{}
	for rows.Next() {
		var k, v sql.NullString
		if err := rows.Scan(&k, &v); err != nil {
			return notifConfig{}, err
		}
		s[k.String] = v.String
	}
	if err := rows.Err(); err != nil {
		return notifConfig{}, err
	}
	cfg := notifConfig{
		appName:           s["app_name"],
		appURL:            os.Getenv("APP_URL"),
		discordEnabled:    s["notifications_discord_enabled"] == "true",
		telegramEnabled:   s["notifications_telegram_enabled"] == "true",
		userConfigEnabled: s["notifications_user_config_enabled"] == "true",
	}
	if cfg.appName == "" {
		cfg.appName = "Struxa"
	}
	if cfg.appURL == "" {
		cfg.appURL = s["app_url"]
	}
	if v := s["notifications_discord_webhook_url"]; v != "" {
		if plain, err := crypto.DecryptToken(v); err == nil {
			cfg.discordURL = plain
		}
	}
	if v := s["notifications_telegram_bot_token"]; v != "" {
		if plain, err := crypto.DecryptToken(v); err == nil {
			cfg.telegramToken = plain
		}
	}
	cfg.telegramChatID = s["notifications_telegram_chat_id"]
	return cfg, nil
}

type ownerNotifConfig struct {
	discordURL     string
	telegramToken  string
	telegramChatID string
}

func loadOwnerConfig(db *sql.DB, userID string) (ownerNotifConfig, error) {
	var webhook, token, chatID sql.NullString
	err := db.QueryRow(
		`SELECT notification_discord_webhook, notification_telegram_token, notification_telegram_chat_id FROM user WHERE id=?`,
		userID,
	).Scan(&webhook, &token, &chatID)
	if err != nil {
		return ownerNotifConfig{}, err
	}
	var c ownerNotifConfig
	if webhook.Valid && webhook.String != "" {
		if plain, err := crypto.DecryptToken(webhook.String); err == nil {
			c.discordURL = plain
		}
	}
	if token.Valid && token.String != "" {
		if plain, err := crypto.DecryptToken(token.String); err == nil {
			c.telegramToken = plain
		}
	}
	if chatID.Valid {
		c.telegramChatID = chatID.String
	}
	return c, nil
}

type statusNotifyJob struct {
	serverID   string
	serverName string
	uuid       string
	ownerID    string
	state      string
}

var (
	statusNotifyMu   sync.Mutex
	statusNotifyJobs = map[string]statusNotifyJob{}
	statusNotifyWake = make(chan struct{}, 1)
)

func enqueueStatusNotify(job statusNotifyJob) {
	statusNotifyMu.Lock()
	statusNotifyJobs[job.serverID] = job
	statusNotifyMu.Unlock()
	select {
	case statusNotifyWake <- struct{}{}:
	default:
	}
}

func runStatusNotifyDispatcher(db *sql.DB) {
	for range statusNotifyWake {
		for {
			statusNotifyMu.Lock()
			var job statusNotifyJob
			ok := false
			for k, v := range statusNotifyJobs {
				job = v
				delete(statusNotifyJobs, k)
				ok = true
				break
			}
			statusNotifyMu.Unlock()
			if !ok {
				break
			}
			notifyServerStatus(db, job.serverID, job.serverName, job.uuid, job.ownerID, job.state)
		}
	}
}

func notifyServerStatus(db *sql.DB, serverID, serverName, uuid, ownerID, state string) {
	if state != "running" && state != "offline" {
		return
	}
	cfg, err := loadNotifConfig(db)
	if err != nil {
		log.Printf("[status] failed to load notification config: %v", err)
		return
	}
	label := "online"
	title := "Server online"
	if state == "offline" {
		label = "offline"
		title = "Server offline"
	}
	short := uuid
	if len(short) > 8 {
		short = short[:8]
	}
	msg := fmt.Sprintf("[%s] Server %q (%s) is now %s", cfg.appName, serverName, short, label)
	n := notifMessage{
		title:       title,
		subtitle:    fmt.Sprintf("%s is now %s.", serverName, label),
		detail:      fmt.Sprintf("**%s** · %s", serverName, short),
		buttonLabel: "View server",
	}
	if cfg.appURL != "" {
		n.buttonURL = cfg.appURL + "/servers/" + uuid
	}

	if cfg.discordEnabled && cfg.discordURL != "" {
		if err := postDiscord(cfg.discordURL, n); err != nil {
			log.Printf("[status] discord notify failed for server %s: %v", serverID, err)
		}
	}
	if cfg.telegramEnabled && cfg.telegramToken != "" && cfg.telegramChatID != "" {
		if err := postTelegram(cfg.telegramToken, cfg.telegramChatID, msg); err != nil {
			log.Printf("[status] telegram notify failed for server %s: %v", serverID, err)
		}
	}

	if ownerID == "" || !cfg.userConfigEnabled {
		return
	}
	oc, err := loadOwnerConfig(db, ownerID)
	if err != nil {
		log.Printf("[status] failed to load owner notification config for %s: %v", ownerID, err)
		return
	}
	if oc.discordURL != "" {
		if err := postDiscord(oc.discordURL, n); err != nil {
			log.Printf("[status] discord notify failed for owner %s: %v", ownerID, err)
		}
	}
	if oc.telegramToken != "" && oc.telegramChatID != "" {
		if err := postTelegram(oc.telegramToken, oc.telegramChatID, msg); err != nil {
			log.Printf("[status] telegram notify failed for owner %s: %v", ownerID, err)
		}
	}
}

type notifMessage struct {
	title       string
	subtitle    string
	detail      string
	buttonLabel string
	buttonURL   string
}

func postDiscord(url string, v notifMessage) error {
	header := fmt.Sprintf("### ✦ %s\n%s", v.title, v.subtitle)
	inner := []any{}
	if v.buttonURL != "" {
		inner = append(inner,
			map[string]any{
				"type": 9,
				"components": []any{
					map[string]any{"type": 10, "content": header},
				},
				"accessory": map[string]any{"type": 2, "style": 5, "label": v.buttonLabel, "url": v.buttonURL},
			},
			map[string]any{"type": 14, "divider": true, "spacing": 1},
		)
	} else {
		inner = append(inner, map[string]any{"type": 10, "content": header})
	}
	inner = append(inner,
		map[string]any{"type": 10, "content": v.detail},
		map[string]any{"type": 10, "content": fmt.Sprintf("-# <t:%d:R>", time.Now().Unix())},
	)
	payload := map[string]any{
		"flags":            32768,
		"allowed_mentions": map[string]any{"parse": []any{}},
		"components": []any{
			map[string]any{"type": 17, "accent_color": nil, "components": inner},
		},
	}
	data, _ := json.Marshal(payload)
	return httpPostJSON(withComponentsParam(url), data)
}

func withComponentsParam(url string) string {
	if strings.Contains(url, "?") {
		return url + "&with_components=true"
	}
	return url + "?with_components=true"
}

func postTelegram(token, chatID, text string) error {
	body, _ := json.Marshal(map[string]string{"chat_id": chatID, "text": text})
	return httpPostJSON(fmt.Sprintf("https://api.telegram.org/bot%s/sendMessage", token), body)
}

func httpPostJSON(target string, body []byte) error {
	client := &http.Client{Timeout: notifyHTTPTimeout}
	resp, err := client.Post(target, "application/json", bytes.NewReader(body))
	if err != nil {
		var urlErr *url.Error
		if errors.As(err, &urlErr) {
			return fmt.Errorf("post request failed: %w", urlErr.Err)
		}
		return err
	}
	defer resp.Body.Close()
	if resp.StatusCode < 200 || resp.StatusCode > 299 {
		return fmt.Errorf("http %d", resp.StatusCode)
	}
	return nil
}

File: packages/api/src/services/notifications.ts
import { eq } from "drizzle-orm";
import { db } from "@struxa/db";
import { settings } from "@struxa/db";
import { user as userTable } from "@struxa/db";
import { env } from "@struxa/env/server";
import { safeDecrypt } from "../lib/crypto";

export const SENTINEL = "[set]";

const TIMEOUT_MS = 8000;

export function isValidDiscordWebhookUrl(raw: string): boolean {
  let u: URL;
  try {
    u = new URL(raw);
  } catch {
    return false;
  }
  return (
    u.protocol === "https:" &&
    (u.hostname === "discord.com" || u.hostname === "discordapp.com") &&
    u.pathname.startsWith("/api/webhooks/")
  );
}

export function isValidTelegramBotToken(token: string): boolean {
  return /^\d{6,}:[A-Za-z0-9_-]{30,}$/.test(token);
}

export function isValidTelegramChatId(id: string): boolean {
  return /^-?\d{5,}$/.test(id);
}

export async function postDiscord(url: string, payload: object): Promise<void> {
  if (!isValidDiscordWebhookUrl(url)) return;
  const res = await fetch(withComponentsParam(url), {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(payload),
    signal: AbortSignal.timeout(TIMEOUT_MS),
  });
  if (!res.ok) throw new Error(`discord ${res.status}`);
}

function withComponentsParam(url: string): string {
  return url.includes("?") ? `${url}&with_components=true` : `${url}?with_components=true`;
}

export async function postTelegram(token: string, chatId: string, text: string): Promise<void> {
  if (!isValidTelegramBotToken(token) || !isValidTelegramChatId(chatId)) return;
  const res = await fetch(`https://api.telegram.org/bot${token}/sendMessage`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ chat_id: chatId, text }),
    signal: AbortSignal.timeout(TIMEOUT_MS),
  });
  if (!res.ok) throw new Error(`telegram ${res.status}`);
}

async function readSettingsMap(): Promise<Record<string, string>> {
  const rows = await db.select().from(settings);
  return Object.fromEntries(rows.map((r) => [r.key, r.value ?? ""]));
}

export async function getAdminNotificationConfig() {
  const s = await readSettingsMap();
  return {
    appName: s.app_name ?? "Struxa",
    appUrl: env.APP_URL ?? s.app_url ?? "",
    discordEnabled: s.notifications_discord_enabled === "true",
    discordWebhookUrl: s.notifications_discord_webhook_url ? safeDecrypt(s.notifications_discord_webhook_url) : "",
    telegramEnabled: s.notifications_telegram_enabled === "true",
    telegramToken: s.notifications_telegram_bot_token ? safeDecrypt(s.notifications_telegram_bot_token) : "",
    telegramChatId: s.notifications_telegram_chat_id ?? "",
  };
}

export async function getUserNotificationConfig(userId: string) {
  const [s, owner] = await Promise.all([
    readSettingsMap(),
    db.query.user.findFirst({ where: eq(userTable.id, userId) }),
  ]);
  if (s.notifications_user_config_enabled !== "true" || !owner) return null;
  return {
    appName: s.app_name ?? "Struxa",
    appUrl: env.APP_URL ?? s.app_url ?? "",
    discordWebhookUrl: owner.notificationDiscordWebhook ? safeDecrypt(owner.notificationDiscordWebhook) : "",
    telegramToken: owner.notificationTelegramToken ? safeDecrypt(owner.notificationTelegramToken) : "",
    telegramChatId: owner.notificationTelegramChatId ?? "",
  };
}

export function notifyAdmins(kind: string, fields: Record<string, string> = {}): void {
  void (async () => {
    try {
      const cfg = await getAdminNotificationConfig();
      const message = buildMessage(kind, { ...fields, appName: cfg.appName });
      const components = buildDiscordComponents(kind, { ...fields, appName: cfg.appName, appUrl: cfg.appUrl });
      if (cfg.discordEnabled && cfg.discordWebhookUrl) await postDiscord(cfg.discordWebhookUrl, components);
      if (cfg.telegramEnabled && cfg.telegramToken && cfg.telegramChatId) {
        await postTelegram(cfg.telegramToken, cfg.telegramChatId, message);
      }
    } catch (e) {
      console.error("[notifications] admin send failed:", e);
    }
  })();
}

export function notifyServerOwner(ownerId: string, kind: string, fields: Record<string, string> = {}): void {
  if (!ownerId) return;
  void (async () => {
    try {
      const cfg = await getUserNotificationConfig(ownerId);
      if (!cfg) return;
      const message = buildMessage(kind, { ...fields, appName: cfg.appName });
      const components = buildDiscordComponents(kind, { ...fields, appName: cfg.appName, appUrl: cfg.appUrl });
      if (cfg.discordWebhookUrl) await postDiscord(cfg.discordWebhookUrl, components);
      if (cfg.telegramToken && cfg.telegramChatId) {
        await postTelegram(cfg.telegramToken, cfg.telegramChatId, message);
      }
    } catch (e) {
      console.error("[notifications] user send failed:", e);
    }
  })();
}

export function buildMessage(kind: string, fields: Record<string, string>): string {
  const prefix = fields.appName ?? "Struxa";
  switch (kind) {
    case "server-power":
      return `[${prefix}] Server "${fields.serverName}" ${fields.action} by ${fields.actor}`;
    case "backup":
      return `[${prefix}] Backup "${fields.backupName}" of "${fields.serverName}" ${fields.result}`;
    case "server-created":
      return `[${prefix}] Server "${fields.serverName}" created`;
    case "server-deleted":
      return `[${prefix}] Server "${fields.serverName}" deleted`;
    case "user-deleted":
      return `[${prefix}] User ${fields.userEmail} deleted`;
    case "payment":
      return `[${prefix}] Payment ${fields.result}${fields.amount ? `: ${fields.amount} ${fields.currency}` : ""} via ${fields.provider}`;
    case "test":
      return `[${prefix}] Test notification`;
    default:
      return `[${prefix}] ${fields.message ?? ""}`;
  }
}

const DISCORD_COMPONENTS_V2_FLAG = 32768;

export function buildDiscordComponents(kind: string, fields: Record<string, string>) {
  const appUrl = fields.appUrl ?? "";
  let title = "Notification";
  let subtitle = "";
  let detail = "";
  let buttonLabel = "Open panel";
  let buttonUrl: string | null = appUrl || null;

  switch (kind) {
    case "server-power":
      title = "Power action";
      subtitle = `${fields.serverName} was ${fields.action}ed by ${fields.actor}.`;
      detail = `**${fields.serverName}** · ${fields.action} by **${fields.actor}**`;
      buttonLabel = "View server";
      if (appUrl && fields.serverUuid) buttonUrl = `${appUrl}/servers/${fields.serverUuid}`;
      break;
    case "backup": {
      const result = fields.result ?? "";
      title = "Backup update";
      subtitle = `${fields.backupName} — ${result}.`;
      detail = `**${fields.backupName}** · ${fields.serverName}`;
      buttonLabel = "View server";
      if (appUrl && fields.serverUuid) buttonUrl = `${appUrl}/servers/${fields.serverUuid}`;
      break;
    }
    case "server-created":
      title = "Server created";
      subtitle = `${fields.serverName} has been created.`;
      detail = `**${fields.serverName}** · created by **${fields.actor}**`;
      buttonLabel = "View server";
      if (appUrl && fields.serverUuid) buttonUrl = `${appUrl}/servers/${fields.serverUuid}`;
      break;
    case "server-deleted":
      title = "Server deleted";
      subtitle = `${fields.serverName} has been deleted.`;
      detail = `**${fields.serverName}** · deleted by **${fields.actor}**`;
      buttonLabel = "View servers";
      if (appUrl) buttonUrl = `${appUrl}/admin/servers`;
      break;
    case "user-deleted":
      title = "User deleted";
      subtitle = "A user account has been deleted.";
      detail = `**${fields.userEmail}** · deleted by **${fields.actor}**`;
      buttonLabel = "View users";
      if (appUrl) buttonUrl = `${appUrl}/admin/users`;
      break;
    case "payment": {
      const received = fields.result === "received";
      title = received ? "Payment received" : "Payment failed";
      subtitle = received ? "A wallet top-up was received." : "A wallet top-up failed.";
      detail = fields.amount ? `**${fields.amount} ${fields.currency}** · ${fields.provider}` : `**${fields.provider}**`;
      buttonLabel = "View billing";
      if (appUrl) buttonUrl = `${appUrl}/admin/billing`;
      break;
    }
    case "test":
      title = "Test notification";
      subtitle = "Notifications are working.";
      detail = `**${fields.appName}** · Discord`;
      break;
  }

  return componentsPayload({ title, subtitle, detail, buttonLabel, buttonUrl });
}

function componentsPayload(v: { title: string; subtitle: string; detail: string; buttonLabel: string; buttonUrl: string | null }) {
  const header = `### ✦ ${v.title}\n${v.subtitle}`;
  const inner: unknown[] = [];
  if (v.buttonUrl) {
    inner.push({
      type: 9,
      components: [{ type: 10, content: header }],
      accessory: { type: 2, style: 5, label: v.buttonLabel, url: v.buttonUrl },
    });
    inner.push({ type:14, divider: true, spacing: 1 });
  } else {
    inner.push({ type: 10, content: header });
  }
  inner.push({ type: 10, content: v.detail });
  inner.push({ type: 10, content: `-# <t:${Math.floor(Date.now() / 1000)}:R>` });
  return {
    flags: DISCORD_COMPONENTS_V2_FLAG,
    components: [{ type: 17, accent_color: null, components: inner }],
  };
}
